src/layers/TooltipsLayer.ts

```diff
@@ -0,0 +1,120 @@
+import * as d3 from "@/d3";
+import { D3Selection, LayerArgs, LayerType, OptionalLayer, Point, XY, CustomEvents } from "./Layer";
+import { TracesLayer } from "./TracesLayer";
+
+export type TooltipHtmlCallback = (point: Point) => string
+
+export class TooltipsLayer extends OptionalLayer {
+  type = LayerType.Tooltip;
+
+  constructor(public tooltipHtmlCallback: TooltipHtmlCallback) {
+    super();
+  };
+
+  // straight line distance squared
+  private l22 = (coord1: Point, coord2: Point) => {
+    const diffX = coord1.x - coord2.x;
+    const diffY = coord1.y - coord2.y;
+    return diffX * diffX + diffY * diffY;
+  };
+
+  private l22Svg = (svgCoord1: Point, svgCoord2: Point, range: XY<number>) => {
+    // we want the closest point based on pixels (how it looks on the user's screen)
+    // however if we just calculated the straight line distance between two points
+    // without scaling the coords then we would get the closest point if the axes were
+    // to scale (i.e. if x axis is [0, 100] and y axis is [0, 700] then the y axis
+    // would be 7 times bigger than the x axis).
+    //
+    // This is not the case as axes may be squished so we need to scale our coordinates
+    // to work out a distance that reflect pixel distance between mouse and a point in
+    // the data
+    const scaledCoord1: Point = { x: svgCoord1.x / range.x, y: svgCoord1.y / range.y };
+    const scaledCoord2: Point = { x: svgCoord2.x / range.x, y: svgCoord2.y / range.y };
+
+    return this.l22(scaledCoord1, scaledCoord2);
+  };
+
+  private handleMouseMove = (event: d3.ClientPointEvent, layerArgs: LayerArgs, tooltip: D3Selection<HTMLDivElement>, allPoints: Point[], range: XY<number>) => {
+    const [ clientX, clientY ] = d3.pointer(event);
+    const { x: scaleX, y: scaleY } = layerArgs.scaleConfig.linearScales;
+    // convert point from pixel position of mouse to coordinates of the axes
+    // in the plot
+    const coords: Point = { x: scaleX.invert(clientX), y: scaleY.invert(clientY) };
+
+    let minDistance = Infinity;
+    const minPoint: Point = allPoints.reduce((minP, p) => {
+      const distance = this.l22Svg(coords, p, range);
+      if (distance >= minDistance) return minP;
+      minDistance = distance;
+      return p;
+    }, { x: 0, y: 0 });
+
+    // The minimum point we have calculated is relative to the coordinate system
+    // of the svg, however we need absolute coordinates on the page to calculate
```

Comment:
I'm now not really sure what you mean by "the coordinate system of the svg" - in terms of data values in the scales? Or the pixel values within the rendered svg (i.e. not relative to the screen)?